File: zig/405.zig
const std = @import("std");

fn solution(allocator: std.mem.Allocator, num: i32) ![]const u8 {
    var n: u32 = @bitCast(num);

    var temp: std.ArrayList(u8) = .empty;
    defer temp.deinit(allocator);

    const ch = "0123456789abcdef";

    while (n > 0) {
        const nc = ch[@mod(n, 16)];
        try temp.append(allocator, nc);
        n >>= 4;
    }

    for (0..temp.items.len / 2) |idx| {
        const t = temp.items[idx];
        temp.items[idx] = temp.items[temp.items.len - 1 - idx];
        temp.items[temp.items.len - 1 - idx] = t;
    }

    return temp.toOwnedSlice(allocator);
}

const expect = std.testing.expect;

test "test case #1" {
    const allocator = std.testing.allocator;

    const input = 26;
    const expected = "1a";

    const actual = try solution(allocator, input);
    defer allocator.free(actual);

    try expect(std.mem.eql(u8, expected, actual));
}

test "test case #2" {
    const allocator = std.testing.allocator;

    const input = -1;
    const expected = "ffffffff";

    const actual = try solution(allocator, input);
    defer allocator.free(actual);

    try expect(std.mem.eql(u8, expected, actual));
}

File: go/405.go
package main

func toHex(num int) string {
	if num == 0 {
		return "0"
	}

	n := uint32(num)

	mp := map[uint32]byte{
		0:  '0',
		1:  '1',
		2:  '2',
		3:  '3',
		4:  '4',
		5:  '5',
		6:  '6',
		7:  '7',
		8:  '8',
		9:  '9',
		10: 'a',
		11: 'b',
		12: 'c',
		13: 'd',
		14: 'e',
		15: 'f',
	}
	temp := []byte{}

	mask := 0b1111

	for n > 0 {
		val := n & uint32(mask)
		ch := mp[val]

		temp = append([]byte{ch}, temp...)

		n >>= 4
	}

	return string(temp)
}
